linera-sdk/src/test/integration/chain.rs

```diff
@@ -362,12 +384,67 @@ impl ActiveChain {
         panic!("Bytecode not found in the chain it was supposed to be published on");
     }
 
+    /// Registers on this chain an application created on another chain.
+    pub async fn register_application<Abi>(&self, application_id: ApplicationId<Abi>) {
+        if self.needs_application_description(application_id).await {
+            let source_chain = self.validator.get_chain(&application_id.creation.chain_id);
+
+            self.add_block(|block| {
+                block.with_request_for_application(application_id);
+            })
+            .await;
+
+            let request = EffectId {
+                chain_id: self.id(),
+                height: self.tip_height().await,
+                index: 0,
+            };
+
+            source_chain
+                .add_block(|block| {
+                    block.with_incoming_message(request);
+                })
+                .await;
+
+            let response = EffectId {
+                chain_id: source_chain.id(),
+                height: source_chain.tip_height().await,
+                index: 0,
+            };
+
+            self.add_block(|block| {
+                block.with_incoming_message(response);
+            })
+            .await;
+        }
+    }
+
+    /// Checks if the `application_id` is missing from this microchain.
+    async fn needs_application_description<Abi>(&self, application_id: ApplicationId<Abi>) -> bool {
+        let applications = self
+            .validator
+            .worker()
+            .await
+            .load_application_registry(self.id())
+            .await
+            .expect("Failed to load application registry");
+
+        match applications
+            .describe_application(application_id.forget_abi())
+            .await
+        {
+            Ok(_) => false,
+            Err(SystemExecutionError::UnknownApplicationId(_)) => true,
+            Err(_) => panic!("Failed to check known bytecode locations"),
+        }
+    }
+
     /// Executes a `query` on an `application`'s state on this microchain.
     ///
     /// Returns the deserialized `Output` response from the `application`.
-    pub async fn query<Output>(
+    pub async fn query<Abi, Output>(
         &self,
-        application_id: ApplicationId,
+        application_id: ApplicationId<Abi>,
         query: impl AsRef<str>,
     ) -> Output
```

Comment:
We'd like to use `Abi::Query` as input and `Abi::QueryResponse` as return type.
Also, I don't think we should assume anything about the query: some apps may use pure JSON (not GraphQL).

Reply:
Done, thanks!